### ISSUE TITLE
Whether this set of analysis theory is applicable to recurrent neural networks

### ISSUE THREAD

**Issue by 9527-ly** [CalculatedContent/WeightWatcher#145]
I noticed that the current toolkit only supports Conv2D and Dense models. But in practical application and research, we often use time series data to predict. So I want to know whether this theory is applicable to recurrent neural networks, such as LSTM and RNN. Thank you very much.

**Comment by 9527-ly** [comment]
Each layer of LSTMs contains multiple weight matrices. After I took out the weight of LSTMs, I concatenate them together. Some of the code I used to test was as follows：

def analyze_model(model):
    alphas = []
    for l in model.layers:
        print(l)
        if l.get_weights():
            if (isinstance(l, tensorflow.keras.layers.LSTM)):
                W=l.get_weights()[1]
            elif (isinstance(l, tensorflow.keras.layers.ConvLSTM2D)):
                W0 = l.get_weights()[0]
                W1 = l.get_weights()[1]
                W = np.append(W0, W1, axis=2)
            else:
                W = l.get_weights()[0]
            W = reshape_tensor(W)
            M, N = np.min(W.shape), np.max(W.shape)
            if ((N>2) and (M>2)):
                Q=N/M
                svd = TruncatedSVD(n_components=M-1, n_iter=7, random_state=10)
                svd.fit(W)
                sv = svd.singular_values_
                evals_imagenet = sv*sv 
                alpha, D, best =fit_powerlaw(evals_imagenet)
                print(alpha, D, best)
                alphas.append(alpha)
    return alphas

**Comment by 9527-ly** [comment]
I want to know whether I have handled the weight of LSTMs correctly. Can my operation apply to other metrics, like the spectral norm and the MP soft rank.

**Comment by charlesmartin14** [comment]
Adding LSTMs is more complicated. Its unclear to me if just reshaping or stacking the matrices will yield power law distributions

**Comment by 9527-ly** [comment]
> Adding LSTMs is more complicated. Its unclear to me if just reshaping or stacking the matrices will yield power law distributions
> 
> also, i dont have a good set of very well trained models, similar to whats on huggingface, to study how LSTMs behave

Thank you. I once saw a paper named ''On Generalization Bounds of a Family of Recurrent Neural Networks''. The link to this paper is https://arxiv.org/pdf/1910.12947v2.pdf.

 Can you provide me with your email address. If possible, I would like to provide you with some LSTMs models for research.

**Comment by charlesmartin14** [comment]
charlesmartin14@gmail.com
charles@calculationconsulting.com

FYI, weightwatcher is not based on generalization bounds; it uses techniques from statistical mechanics

If you can send me a few models (and maybe the code used to generate them) I can see about adding them to weightwatcher.  LIke a shared google drive would be good.

**Comment by 9527-ly** [comment]
> [charlesmartin14@gmail.com](mailto:charlesmartin14@gmail.com) [charles@calculationconsulting.com](mailto:charles@calculationconsulting.com)
> 
> FYI, weightwatcher is not based on generalization bounds; it uses techniques from statistical mechanics
> 
> If you can send me a few models (and maybe the code used to generate them) I can see about adding them to weightwatcher. LIke a shared google drive would be good.

Thank you @charlesmartin14. In recent days, I will provide you with the tensorflow model according to your needs.

**Comment by 9527-ly** [comment]
@charlesmartin14. I uploaded the file. Due to the limitation of the upload file size, I have compressed the input feature file. You can get the file from the link: https://github.com/9527-ly/lstms.  You can also get it from the email I sent you.

**Comment by charlesmartin14** [comment]
This will take some time
We have started a channel on discord to discuss

**Comment by 9527-ly** [comment]
> This will take some time We have started a channel on discord to discuss
Thank you very much for your attention to this work. I look forward to its realization

**Comment by 9527-ly** [comment]
> This will take some time We have started a channel on discord to discuss

Thank you very much for your attention to this work. I look forward to its realization
